### PR TITLE
CI Add CPython 3.13 free-threaded test

### PR DESCRIPTION
**Pull request by lesteve**
Giving it a go to CPython 3.13 free-threaded which is in beta and scheduled to be released in October 2024. Context: https://github.com/scikit-learn/scikit-learn/issues/28978 and more generally https://github.com/Quansight-Labs/free-threaded-compatibility.

Locally and in a quick Github action workflow (see [latest runs](https://github.com/lesteve/joblib/actions?query=branch%3Acpython-free-threaded)) there are the following test failures. Note: this is easier to write a Github Action with the action `deadsnakes/action` which allow to have an easily working Python 3.13 beta.

### DeprecationWarning about forking multi-threaded processes

Doing a `Parallel` with loky followed by a `Parallel` by multiprocessing yields a `DeprecationWarning` because I guess `loky` creates thread and doing a fork of a multi-threaded warning may lead to issues. Maybe we should swallow warning in the test? Not sure if there are use cases where using both loky and multiprocessing and what to do in this case.

```py
from joblib import delayed, Parallel


def square(x):
    return x**2


Parallel(n_jobs=2, backend="loky")(delayed(square)(x) for x in range(3))

Parallel(n_jobs=2, backend="multiprocessing")(delayed(square)(x) for x in range(3))
```

produces this warning:
```
/opt/python-free-threading-debug/lib/python3.13/multiprocessing/popen_fork.py:67: DeprecationWarning: This process (pid=802913) is multi-threaded, use of fork() may lead to deadlocks in the child.
  self.pid = os.fork()
```

```
/opt/python-free-threading-debug/lib/python3.13/multiprocessing/popen_fork.py:67: DeprecationWarning: This process (pid=802913) is multi-threaded, use of fork() may lead to deadlocks in the child.
  self.pid = os.fork()
```

The failures are:
```
FAILED joblib/test/test_parallel.py::test_main_thread_renamed_no_warning[None] - assert 2 == 0
FAILED joblib/test/test_parallel.py::test_main_thread_renamed_no_warning[multiprocessing] - assert 2 == 0
FAILED joblib/test/test_parallel.py::test_main_thread_renamed_no_warning[backend7] - assert 2 == 0
```

There are some warnings like this where the test expects no warning:
```
DeprecationWarning('This process (pid=695921) is multi-threaded, use of fork() may lead to deadlocks in the child.')
```

### Fixed: Changes in `locals()` after `exec`

I fixed it by tweaking the test and passing a dict to the `exec` `locals` argument. I think this still tests interactively defined functions.

```
FAILED joblib/test/test_memory.py::test_parallel_call_cached_function_defined_in_jupyter[True] - KeyError: 'f'
FAILED joblib/test/test_memory.py::test_parallel_call_cached_function_defined_in_jupyter[False] - KeyError: 'f'
```

This happens in the test that is trying to test an interactive function in Jupyter by using `exec` to define a function and expecting `locals()` to be udpated. This is likely related to CPython 3.13 (not free-threaded related). 

Probably related:
- https://github.com/python/cpython/issues/118888
- https://docs.python.org/3.13/whatsnew/3.13.html#summary-release-highlights 
  > [PEP 667](https://peps.python.org/pep-0667/): The [locals()](https://docs.python.org/3.13/library/functions.html#locals) builtin now has [defined semantics](https://docs.python.org/3.13/whatsnew/3.13.html#whatsnew313-locals-semantics) when mutating the returned mapping. Python debuggers and similar tools may now more reliably update local variables in optimized scopes even during concurrent code execution.

### CODE REVIEW

**Comment by rgommers** [comment]
I was just looking at the `multiprocessing` fork method deprecation warning, since they also show up in SciPy tests. Given that the default method will change in Python 3.14 (https://docs.python.org/3.13/whatsnew/3.13.html#pending-removal-in-python-3-14), did you consider always specifying the start method, and/or defaulting to `forkserver`? It's actually pretty unclear to me (even after reading https://github.com/python/cpython/issues/84559#issuecomment-1416278935 and other comments in that issue) whether we're supposed to fix it in libraries or let users deal with it themselves by setting `multiprocessing.set_start_method` in case they don't want the defaults to switch from under them in 3.14

**Comment by lesteve** [comment]
Thanks for the pointers, so indeed the change is that on Linux in CPython 3.14, the default multiprocessing start method will change from 'fork' to a better default (I haven't read enough to figure out what the new default is :sweat_smile:)

To quote the relevant section for more precision:

> [multiprocessing](https://docs.python.org/3.13/library/multiprocessing.html#module-multiprocessing): The default start method will change to a safer one on Linux, BSDs, and other non-macOS POSIX platforms where 'fork' is currently the default ([gh-84559](https://github.com/python/cpython/issues/84559)). Adding a runtime warning about this was deemed too disruptive as the majority of code is not expected to care. Use the [get_context()](https://docs.python.org/3.13/library/multiprocessing.html#multiprocessing.get_context) or [set_start_method()](https://docs.python.org/3.13/library/multiprocessing.html#multiprocessing.set_start_method) APIs to explicitly specify when your code requires 'fork'. See [Contexts and start methods](https://docs.python.org/3.13/library/multiprocessing.html#multiprocessing-start-methods).

In joblib `'multiprocessing'` backend case, I am not sure either what we should do. I see broadly three directions:
1. use multiprocessing with `fork` i.e. the default until Python 3.13, at least code that was working with joblib before will still work the same, for example interactive functions (i.e. defined in `__main__`).
2. let the warning through and add some documentation to tell users to set it explicitly the context depending on their use case.
3. swallow the warning because we are expecting most users are happy with the default start method even if it changes in Python 3.14. Document the caveat when Python 3.14 is released to say that if you had code that was working before, here is what you need to keep it working.

I would lean slightly towards 3. maybe? Probably not 1. because my understanding is that joblib `'multiprocessing'` backend should respect the start method if set by the user.

To be honest I am not that sure how much the joblib `'multiprocessing'` backend is used ... but maybe a significant portion of these users do it because they rely on the `fork` behaviour, hard to tell ...

cc @tomMoral and @ogrisel for informed opinions on this topic

**Comment by rgommers** [comment]
> I haven't read enough to figure out what the new default is

My understanding is: very likely `'forkserver'`, but the devs reserved the right to change that decision during the 3.14 alpha phase.

**Comment by ogrisel** [comment]
> Doing a Parallel with loky followed by a Parallel by multiprocessing yields a DeprecationWarning because I guess loky creates thread and doing a fork of a multi-threaded warning may lead to issues. Maybe we should swallow warning in the test? Not sure if there are use cases where using both loky and multiprocessing and what to do in this case.

I agree we might want the swallow this warning in our test. This is quite a synthetic case that should not happen much in practice, so we can ignore and leave the deprecation warning go through in joblib.

**Comment by ogrisel** [comment]
> did you consider always specifying the start method, and/or defaulting to forkserver? It's actually pretty unclear to me (even after reading https://github.com/python/cpython/issues/84559#issuecomment-1416278935 and other comments in that issue) whether we're supposed to fix it in libraries or let users deal with it themselves by setting multiprocessing.set_start_method in case they don't want the defaults to switch from under them in 3.14

I think I would let the `multiprocessing` users chose their own start method. I would not change the default because:

- it would be a behavioral breaking change for people who rely on the semantics of the `fork` start method;
- joblib's default `loky` backend is "better" than `multiprocessing` + `forkserver` because it allows to call functions defined interactively in the `__main__` module or as local variable under the scope of another function. So I would not push users towards a solution that does not allow that.

Once CPython 3.14 has settled down on its plan, we can adapt joblib accordingly to adapt the tests for CPython 3.14 and later but I would not change joblib's default behavior for earlier Python versions.

**Comment by ogrisel** [comment]
> 3. swallow the warning because we are expecting most users are happy with the default start method even if it changes in Python 3.14. Document the caveat when Python 3.14 is released to say that if you had code that was working before, here is what you need to keep it working.

+1. I just assume that "swallow the warning" is only done in when running the tests, not in general in joblib.

**Review by ogrisel — COMMENTED** [review]
Other than the warning problem and the following, LGTM. Thanks Loïc!

File: .github/workflows/cpython-free-threaded.yml [hunk 1]
```diff
@@ -0,0 +1,27 @@
...
+      - run: |
+          python --version
+          which python
+          python -c "import sysconfig; print(sysconfig.get_config_var('Py_GIL_DISABLED'))"
+          python -c "import sys; print(sys._is_gil_enabled())"
+
+      - run: |
+          # Need pip 24.1b1 to find free-threading wheel
+          pip install -U --pre pip
+          pip install pytest pytest-timeout pytest-asyncio==0.21.1 threadpoolctl
```
[the remark this answers is not in the document]
Comment (lesteve) [reply .github/workflows/cpython-free-threaded.yml:22]: This is copy and pasted from above and I don't know the reason :wink:

[the remark this answers is not in the document]
Comment (lesteve) [reply .github/workflows/cpython-free-threaded.yml]: Actually I have removed this since. Free-threaded Python is tested on Azure-pipelines

**Comment by lesteve** [comment]
OK, CI is green and this is ready for another review round!

I have mostly:
- added a free-threaded Python 3.13 is in Azure rather than Github action
- ignored the fork of multi-threaded process warning in the test

File: joblib/test/common.py [hunk 2]
```diff
@@ -14,6 +15,8 @@
     lz4 = None
 
 IS_PYPY = hasattr(sys, "pypy_version_info")
+IS_GIL_DISABLED = (
+    sysconfig.get_config_var("Py_GIL_DISABLED") and not sys._is_gil_enabled())
```
Comment (tomMoral) [inline joblib/test/common.py:19]: nitpick
```suggestion
    sysconfig.get_config_var("Py_GIL_DISABLED") and not sys._is_gil_enabled()
)
```

**Review by tomMoral — APPROVED** [review]
This is surprisingly a small diff 😅

thanks for having anticipated so much with support for nogil @lesteve !

**Comment by tomMoral** [comment]
Thanks @lesteve !

**Comment by lesteve** [comment]
> This is surprisingly a small diff 😅

I guess the issue are more likely to happen when using free-threaded + joblib multithreading backend + some other library.

Next thing I want to do is run the joblib test suite with default backend switched to threading. I think locally I have seen some errors but I need to investigate more.

After this trying to run the scikit-learn test suite with default backend switched to threading.

**Pull request merged**